src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -684,19 +685,22 @@ internal sealed override bool IsInterface
             }
         }
 
-        internal override ManagedKind ManagedKind
+        internal override ManagedKind GetManagedKind(ref HashSet<DiagnosticInfo> useSiteDiagnostics)
         {
-            get
+            var managedKind = _flags.ManagedKind;
+            if (managedKind == ManagedKind.Unknown)
             {
-                var managedKind = _flags.ManagedKind;
-                if (managedKind == ManagedKind.Unknown)
-                {
-                    var baseKind = base.ManagedKind;
-                    _flags.SetManagedKind(baseKind);
-                    return baseKind;
-                }
-                return managedKind;
+                managedKind = base.GetManagedKind(ref _managedKindUseSiteDiagnostics);
+                _flags.SetManagedKind(managedKind);
             }
+
+            if (_managedKindUseSiteDiagnostics is object)
```

Comment:
>_managedKindUseSiteDiagnostics [](start = 16, length = 30)

Should we use volatile read here?

Reply:
After changing over to ImmutableArray and initializing with InterlockedExchange I do not think a volatile read is necessary.

Reply:
> After changing over to ImmutableArray and initializing with InterlockedExchange I do not think a volatile read is necessary.

Is something like this possible: 
1. the reads are reordered and first we read ```_managedKindUseSiteDiagnostics``` (default at the moment)
2. then the thread is suspended and some other thread completes all the work
3. the thread is resumed and we read ```managedKind``` from flags.
4. We end up here with default array.

One way to avoid this is to do a volatile read here, or to check both the ```managedKind``` and ```_managedKindUseSiteDiagnostics.IsDefault``` in the if above.

---
In reply to: [458985070](https://github.com/dotnet/roslyn/pull/46114#discussion_r458985070) [](ancestors = 458985070)

Reply:
I do not see how the read of `_managedKindUseSiteDiagnostics` on [line 700](https://github.com/dotnet/roslyn/blob/1986ec4677fa82c499f412a232df296339d4a4cb/src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs#L700) could be reordered to occur before the read of `_flags` via [line 690](https://github.com/dotnet/roslyn/blob/1986ec4677fa82c499f412a232df296339d4a4cb/src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs#L690) because this method may write to `_managedKindUseSiteDiagnostics` in-between in a single-threaded context. However I am far from an expert on the CLR memory model, so I'm happy to insert the volatile read and remove any doubt.

Reply:
I wasn't able to find API that can volatile read an immutable array, but it seems like inserting a `Thread.MemoryBarrier()` call would have a similar effect.